### PR TITLE
Rebuild Support for custom Markdown on Configuration page

### PR DESCRIPTION
**Pull request by montogeek**
This PR add support for custom Markdown in the Configuration page: https://webpack.js.org/configuration/, click the chevrons in the huge code example.

It is using mdx-loader to convert Markdown to React components, once in React, we can do pretty much anything with them.

That is just an example, some other ideas I have:
1. Show that card to the right, outside of the snippet, it would automatically update the content when you hover your mouse on each option, the card will follow your scroll, this won’t be accessible because text won’t exist until it is show.
2. Just fill the page with those cards (accessible by default), these cards would be to the right of the snippet, one problem is space, some options are too large.
3. Make it an accordion that would toggle below the option, but content would be accessible and with better style to made it really _pop_.

It is very important the accessible aspect, we have to make sure the entire doc is searchable using (Cmd|Ctrl)+F, it is a common interaction when looking for documentation.

You can see an demo of how it works here: https://gfycat.com/ArcticAdmiredAntipodesgreenparakeet

### CODE REVIEW

File: jsLinks.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+module.exports = jslinks;
```
Comment (montogeek) [inline jsLinks.js]: This file would be moved to an standalone package.

File: package.json [hunk 2]
```diff
@@ -48,11 +48,6 @@
     "serve": "npm run build && sirv start ./dist --port 4000",
     "deploy": "gh-pages -d dist"
   },
-  "husky": {
```
Comment (montogeek) [inline package.json]: Will revert this change

File: package.json [hunk 3]
```diff
@@ -126,6 +123,11 @@
     "tap-spot": "^1.1.1",
     "through2": "^2.0.3",
     "uglifyjs-webpack-plugin": "^1.1.6",
+    "unist-util-find": "1.0.1",
```
Comment (montogeek) [inline package.json]: These dependencies would be moved to an standalone package.

File: prism.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
+const util = require('util');
```
Comment (montogeek) [inline prism.js]: This file would be moved to an standalone package.

File: src/components/Configuration/Configuration.jsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import React from 'react';
+import { Modes, Entry } from './components';
+
+const components = {
```
Comment (montogeek) [inline src/components/Configuration/Configuration.jsx]: This is actually not necessary, will clean up, only `link` is necessary.

File: src/components/Configuration/components.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        containerClassName={'shadow'}
+        content={<Card body={content} />}
+      >
+        <span className={`code-details-summary-span ${className}`} onClick={() => this.setState({ open: !this.state.open })}>{summary}</span>
+      </Popover>
+    );
+  }
+}
+
+export const Entry = props => {
```
Comment (montogeek) [inline src/components/Configuration/components.js]: Will remove

File: src/components/Page/Page.jsx [hunk 7]
```diff
@@ -28,62 +29,71 @@ class Page extends React.Component {
...
 
     const { contentLoaded } = this.state;
     const loadRelated = contentLoaded && related && related.length !== 0;
     const loadContributors = contentLoaded && contributors && contributors.length !== 0;
 
+    const { content } = this.state;
+
+    let contentRender;
+
+    if (typeof content === 'function') {
```
Comment (montogeek) [inline src/components/Page/Page.jsx]: mdx loader returns a function.

File: prism.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+remark()
+  .use(parse)
+  .use(attacher)
+  .use(htmlRemark)
+  .process(file, function(err, file) {
+    if (err) throw err;
+    if (file) {
+      fs.writeFileSync(
+        './index.html',
+        `<link rel="stylesheet" href="https://webpack.js.org/142727ee4bde248bb419.css" />
```
Comment (EugeneHlushko) [inline prism.js]: ?

File: prism.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+function attacher({ include, exclude } = {}) {
+  return transformer;
+
+  function transformer(tree, file) {
+    visit(tree, 'code', visitor);
+
+    function visitor(node) {
+      const { lang } = node;
+
+      if (!lang || (include && !~include.indexOf(lang)) || (exclude && ~exclude.indexOf(lang))) {
```
Comment (EugeneHlushko) [inline prism.js]: cant we just use `.includes` instead of `.indexOf` trickery?

File: src/components/Configuration/Configuration.scss [hunk 10]
```diff
@@ -0,0 +1,10 @@
+.shadow {
+  overflow: visible;
+  border-radius: 4px;
+  box-shadow: -1px 1px 10px 0 rgba(255, 255, 255, 0.44);
+}
+
+.inline {
+  padding-right: 15px !important;
```
Comment (EugeneHlushko) [inline src/components/Configuration/Configuration.scss]: why important tags? maybe just add more specificity? or are we overriding inline styles here

File: src/components/Configuration/components.js [hunk 11]
```diff
@@ -0,0 +1,82 @@
+import React from 'react';
+import Popover from 'react-tiny-popover';
+import './Configuration.scss';
+
+const isFirstChild = child => typeof child === 'string' && child !== ' ';
```
Comment (EugeneHlushko) [inline src/components/Configuration/components.js]: is it possible that `child === ''` ?

File: src/components/Configuration/components.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+      content: null
+    };
+  }
+
+  componentDidMount() {
+    const { children, url } = this.props;
+
+    const closeDefault = children.findIndex(child => {
+      if (React.isValidElement(child)) {
+        return child.props.props.className.includes('tag') && child.props.children.length === 4;
```
Comment (EugeneHlushko) [inline src/components/Configuration/components.js]: how did we get `.props.props`? maybe worth renaming the latter if its not reacts `props`

File: src/components/Configuration/components.js [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+  render() {
+    const { open, summary, content } = this.state;
+    const className = open ? 'open' : '';
+    return (
+      <Popover
+        isOpen={open}
+        position={['right', 'top']}
+        disableReposition
+        padding={0}
+        onClickOutside={() => this.setState({ open: false })}
```
Comment (EugeneHlushko) [inline src/components/Configuration/components.js]: i would suggest moving such stuff into named class functions, could be:
```
clickOutsideHandler = () => {
  this.setState({ open: false })}
}
```

File: src/components/Configuration/components.js [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        containerClassName={'shadow'}
+        content={<Card body={content} />}
+      >
+        <span className={`code-details-summary-span ${className}`} onClick={() => this.setState({ open: !this.state.open })}>{summary}</span>
```
Comment (EugeneHlushko) [inline src/components/Configuration/components.js]: same here, maybe thats only my preference but having those named gives better error understanding if it happens on prod for example

File: src/content/configuration/index.mdx [hunk 15]
```diff
@@ -0,0 +1,455 @@
+---
```
Comment (EugeneHlushko) [inline src/content/configuration/index.mdx]: i wonder why this shows as a diff and not as a move?

File: prism.js (hunk 8 above)
Comment (montogeek) [reply prism.js]: Read my comments: https://github.com/webpack/webpack.js.org/pull/2564#discussion_r221837410

File: prism.js (hunk 9 above)
Comment (montogeek) [reply prism.js]: Yep

File: src/components/Configuration/Configuration.scss (hunk 10 above)
Comment (montogeek) [reply src/components/Configuration/Configuration.scss]: CSS need reworking, it was to add more specificity

File: src/components/Configuration/components.js (hunk 11 above)
Comment (montogeek) [reply src/components/Configuration/components.js]: Don't think so

File: src/components/Configuration/components.js (hunk 12 above)
Comment (montogeek) [reply src/components/Configuration/components.js]: I don't think so, that is the convention with remark ecosystem.
Will check

File: src/components/Configuration/components.js (hunk 13 above)
Comment (montogeek) [reply src/components/Configuration/components.js]: Yep

File: src/components/Configuration/components.js (hunk 14 above)
Comment (montogeek) [reply src/components/Configuration/components.js]: Yep

File: src/content/configuration/index.mdx (hunk 15 above)
Comment (montogeek) [reply src/content/configuration/index.mdx]: GitHub says the file was removed, I don't remember what I did exactly

**Comment by montogeek** [comment]
> merge the bloody rebuild into master 👯 👯 👯 👨‍🎨 👍
Me too, me too

**Comment by montogeek** [comment]
@EugeneHlushko Please take a look, once this is merged, we can merge `rebuild`!

File: src/components/Configuration/components.js [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    this.setState({ open: !this.state.open });
+  };
+
+  render() {
+    const { open, summary, content } = this.state;
+    const className = open ? "open" : "";
+    return (
+      <Popover
+        isOpen={open}
+        position={["right", "top"]}
```
Comment (EugeneHlushko) [inline src/components/Configuration/components.js]: what if we make this resolution agnostic? e.g. for narrow screens just make it fixed at bottom and limit its vertical space?
screen shot 2018-12-17 at 6 04 36 pm

File: webpack.common.js [hunk 17]
```diff
@@ -1,59 +1,103 @@
...
   module: {
     rules: [
+      {
+        test: /\.mdx$/,
+        use: [
+          "babel-loader",
+          {
+            loader: "@mdx-js/loader",
+            options: {
+              mdPlugins: [
```
[the remark this answers is not in the document]
Comment (montogeek) [reply webpack.common.js]: Yep

File: webpack.common.js [hunk 18]
```diff
@@ -1,59 +1,103 @@
-const fs = require('fs');
-const path = require('path');
-const webpack = require('webpack');
-const CopyWebpackPlugin = require('copy-webpack-plugin');
-const ExtractTextPlugin = require('extract-text-webpack-plugin');
+const fs = require("fs");
```
[the remark this answers is not in the document]
Comment (montogeek) [reply webpack.common.js]: Yep

File: src/components/Page/Page.jsx [hunk 19]
```diff
@@ -28,62 +29,71 @@ class Page extends React.Component {
...
 
     const { contentLoaded } = this.state;
     const loadRelated = contentLoaded && related && related.length !== 0;
     const loadContributors = contentLoaded && contributors && contributors.length !== 0;
 
+    const { content } = this.state;
+
+    let contentRender;
+
+    if (typeof content === "function") {
```
[the remark this answers is not in the document]
Comment (montogeek) [reply src/components/Page/Page.jsx]: Yep

File: src/components/Page/Page.jsx [hunk 20]
```diff
@@ -1,14 +1,15 @@
 // Import External Dependencies
-import React from 'react';
+import React from "react";
```
[the remark this answers is not in the document]
Comment (montogeek) [reply src/components/Page/Page.jsx]: Yep

File: src/components/Configuration/Configuration.jsx [hunk 21]
```diff
@@ -0,0 +1,94 @@
+import React from "react";
+import ReactMarkdown from "react-markdown";
+import { Details } from "./components";
+
+const components = {
```
Comment (EugeneHlushko) [inline src/components/Configuration/Configuration.jsx]: Maybe go with 
```suggestion
const detailComponentsList = ['link', 'mode', 'entry', 'filename', 'publicPath', 'advancedOutput', 'expert', 'advancedModule', 'alias', 'advancedResolve', 'hints', 'devtool', 'target', 'externals', 'stats', 'advanced', 'libraryTarget'];
const detailsComponent = (children, props) => {
  return <Details children={children} {...props} />;
}
```

File: src/components/Configuration/Configuration.jsx [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+  libraryTarget: (children, props) => {
+    return <Details children={children} {...props} />;
+  }
+};
+
+const Pre = props => {
+  const newChildren = React.Children.map(props.children.props.children, child => {
+    if (React.isValidElement(child)) {
+      if (child.props.props.className.includes("keyword")) {
+        if (!components[child.props.props.componentname]) return child;
```
Comment (EugeneHlushko) [inline src/components/Configuration/Configuration.jsx]:
```suggestion
        if (!detailComponentsList.includes[child.props.props.componentname]) return child;
```

File: src/components/Configuration/Configuration.jsx [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+
+        return components[child.props.props.componentname](
```
Comment (EugeneHlushko) [inline src/components/Configuration/Configuration.jsx]:
```suggestion
        return detailsComponent(
```

File: src/components/Configuration/components.js [hunk 24]
```diff
@@ -0,0 +1,98 @@
+import React from 'react';
+import Popover from 'react-tiny-popover';
+import './Configuration.scss';
+import { timeout } from 'q';
+
+const DEFAULT_CHILDREN_SIZE = 4;
+
+const isFirstChild = child => typeof child === 'string' && child !== ' ';
```
Comment (EugeneHlushko) [inline src/components/Configuration/components.js]: I still dont get why we dont do `child.trim() !== ''` instead of doing check for a space, but ok :)

File: src/components/Configuration/Configuration.jsx [hunk 25]
```diff
@@ -0,0 +1,97 @@
+import React from "react";
+import ReactMarkdown from "react-markdown";
+import { Details } from "./components";
+
+const detailComponentsList = ['link', 'mode', 'entry', 'filename', 'publicPath', 'advancedOutput', 'expert', 'advancedModule', 'alias', 'advancedResolve', 'hints', 'devtool', 'target', 'externals', 'stats', 'advanced', 'libraryTarget'];
+const detailsComponent = (children, props) => {
+  return <Details children={children} {...props} />;
+}
+  link: (children, props) => {
```
[the remark this answers is not in the document]
Comment (montogeek) [reply src/components/Configuration/Configuration.jsx]: Not worries, will do it tomorrow

[the remark this answers is not in the document]
Comment (EugeneHlushko) [reply src/components/Configuration/Configuration.jsx]: nice :) thanks man

File: src/components/Configuration/Configuration.jsx (hunk 22 above)
Comment (montogeek) [reply src/components/Configuration/Configuration.jsx]: This doesn't work, I lost 5 minutes debugging it, JS just run `obj.includes["key"]` just returns `undefined` :D

Comment (EugeneHlushko) [reply src/components/Configuration/Configuration.jsx]: lol :DDDD my bad was editing in github

**Comment by EugeneHlushko** [comment]
testing this
